### PR TITLE
Improve Li thresholding

### LINKED ISSUE #2337: threshold_li fails for "zeros" input images due to undefined `threshold`
## Description

When running `skimage.filters.threshold_li`, if the input image is all 0, the 
`tolerance` is set to zero causing the while loop never to execute and a somewhat cryptic error is
encountered because `threshold` never gets set. 

A typical traceback is 

```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    threshold_li(np.zeros((100,100)))
  File "./skimage/filters/thresholding.py", line 509, in threshold_li
    return threshold + immin
UnboundLocalError: local variable 'threshold' referenced before assignment
```

I recognise that this is a special case, but I just wanted to raise this issue as a way of
asking 

> do we want this somewhat cryptic error, or should we catch this 
> using a check on the input image"?  
## Way to reproduce

All that's needed to reproduce this (Python 3.4.1, skimage 0.13dev) is: 

```
import numpy as np
from skimage.filters import threshold_li
threshold_li(np.zeros((100,100)))
```

### PR DESCRIPTION
**Pull request by jni**
- images with np.nan and np.inf values are now processed correctly
- `temp` value becoming negative was impossible, and this line [was not
  covered](https://codecov.io/gh/scikit-image/scikit-image/src/86bbed0937aef504cc1a9837c1a21271cce5bef4/skimage/filters/thresholding.py), so removed it
- code contained numerous fixes for integer offset errors. Since we
  deal with floating point, these are unnecessary. The code is now
  much simpler
- it is also more correct: the tolerance offset that was applied
  previously in fact was giving an incorrect threshold. I verify this
  with a new private function, _cross_entropy, which estimates the
  cross-entropy between foreground and background as defined in the
  original paper.

The only fly in this sweet ointment is that for *one* of our test
images, coins, the threshold found by the iterative method is one off
(or rather 0.something off, but this happens to fall on either side of an
integer) from the optimal threshold. I believe this is an unavoidable
side effect of images being discretized into integer bins.

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] ~~Gallery example in `./doc/examples` (new features only)~~
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
Fixes #2337 

## For reviewers
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/filters/thresholding.py [hunk 1]
```diff
@@ -459,14 +459,67 @@ def threshold_isodata(image, nbins=256, return_all=False):
         return thresholds[0]
 
 
-def threshold_li(image):
-    """Return threshold value based on adaptation of Li's Minimum Cross Entropy method.
+def _cross_entropy(image, threshold, bins=np.arange(-0.5, 255.51, 1)):
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Is the default value intended?

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -459,14 +459,67 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+
+def threshold_li(image, *, tolerance=None):
+    """Compute threshold value by Li's iterative Minimum Cross Entropy method.
+
+    Parameters
+    ----------
+    image : ndarray
         Input image.
 
+    tolerance : float
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: , optional.

File: skimage/filters/thresholding.py [hunk 3]
```diff
@@ -459,14 +459,67 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    m0b = np.sum(histogram[t:])
+    m1a = np.sum(histogram[:t] * bin_centers[:t])  # 1st moment, background
+    m1b = np.sum(histogram[t:] * bin_centers[t:])
+    mua = m1a / m0a  # mean value, background
+    mub = m1b / m0b
+    nu = -m1a * np.log(mua) - m1b * np.log(mub)
+    return nu
+
+
+def threshold_li(image, *, tolerance=None):
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: This signature looks weird

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (jni) [reply skimage/filters/thresholding.py]: Yes. Unfortunately you get some nasty aliasing artifacts if you just write 255 or 256...

File: skimage/filters/thresholding.py (hunk 3 above)
Comment (jni) [reply skimage/filters/thresholding.py]: it is a Python 3 keyword-only argument. https://www.python.org/dev/peps/pep-3102/

**Comment by jni** [comment]
@sciunto thanks for the quick review! I've addressed your docstring comments. For the default value, it's a bit weird, but since it's correct and it's a private function, I figure it can stay like that. I could also set it to None and compute it inside the function. I'm ok with either option, depending on reviewer preferences. =)

File: skimage/filters/thresholding.py (hunk 3 above)
Comment (sciunto) [reply skimage/filters/thresholding.py]: oh, yes, I forgot that one. However, don't we want to backport this? you said that the current version can return incorrect thresholds, which is embarrassing.

Comment (jni) [reply skimage/filters/thresholding.py]: Yes but my philosophy with backports is to code as we want to code (ie in Python 3 if it fits), and then manually fix backports if necessary. The whole idea of dropping support for 2.7 on master is to stop being hamstrung by the need to support old syntax.

In this specific case, we never supported tolerance before so we can hardcode it anyway.

Comment (sciunto) [reply skimage/filters/thresholding.py]: ok :)

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (sciunto) [reply skimage/filters/thresholding.py]: I would prefer to apply your alternative solution, ie, specifying this range before the function call instead as a default value.

File: skimage/filters/thresholding.py [hunk 4]
```diff
@@ -459,14 +459,67 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    """Compute threshold value by Li's iterative Minimum Cross Entropy method.
+
+    Parameters
+    ----------
+    image : ndarray
         Input image.
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: blank line?

File: skimage/filters/thresholding.py [hunk 5]
```diff
@@ -494,44 +547,47 @@ def threshold_li(image):
...
 
     # Stop the iterations when the difference between the
     # new and old threshold values is less than the tolerance
-    while abs(new_thresh - old_thresh) > tolerance:
-        old_thresh = new_thresh
-        threshold = old_thresh + tolerance   # range
-        # Calculate the means of background and object pixels
-        mean_back = image[image <= threshold].mean()
-        mean_obj = image[image > threshold].mean()
+    while abs(t_next - t_curr) > tolerance:
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Naive question. Do we have the warranty that it converges in any case?

**Comment by sciunto** [comment]
I left few comments. Then, it'll be ready to go.

File: skimage/filters/thresholding.py (hunk 5 above)
Comment (jni) [reply skimage/filters/thresholding.py]: @sciunto yes, the paper proves this, although I'll admit that I didn't follow the proof in detail. =)

**Comment by jni** [comment]
@sciunto what do you think of the current solution?

**Comment by sciunto** [comment]
Yeap, that's better. Sorry to bother you again. Could you had the argument you said here to give the reason of that choice? Then, I merge. Promised :)

**Comment by sciunto** [comment]
Merged :)

**Comment by jni** [comment]
@meeseeksdev backport to v0.14.x

**Comment by jni** [comment]
I should have known. LOL

**Comment by sciunto** [comment]
:D

**Pull request merged**